File: 104.maximum-depth-of-binary-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

int max(int x, int y) {
    if (x > y)
        return x;
    return y;
}

/* Approach: DFS + Recursion, Complexity:  O(n), O(n) */
int maxDepth(struct TreeNode* root){
    /* If there is node, the path below haas no depth. */
    if (!root)
        return 0;

    /* Get max length path below a node by selecting the larger of the left and 
     * right paths. Add 1 to account for the node itself. */
    return max(maxDepth(root->left), maxDepth(root->right)) + 1;
}

File: 104.maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Approach: DFS + Stack, Complexity: O(n), O(n)
# Tip: Store nodes in a stack as a path is traversed. Max path in the tree = Max size 
#      of stack during traversal of tree.

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        maxdepth = 0
        stack, top = [], -1
        stack.append([root, 0])

        while stack:
            # Every stack entry holds a node and its state. State determine if
            # its children are visited yet.
            topnode, state = stack[top]

            # Handle empty tree and nulls added due to absense of node.
            if not topnode:
                stack.pop()
                continue

            # Stack size increases as a path is traversed.
            maxdepth = max(maxdepth, len(stack))

            if state == 0:
                stack[top][1] += 1
                stack.append([topnode.left, 0])
            elif state == 1:
                stack[top][1] += 1
                stack.append([topnode.right, 0])
            else:
                stack.pop()

        return maxdepth
